explainer.md

```diff
@@ -0,0 +1,273 @@
+mediaSource.onsourceopen = (e) => {
+  const sourceBuffer = mediaSource.addSourceBuffer(mime);
+  // App-specific request
+  const mediaRequest = Uint8Array.from([1, 2, 3, 4]);
+  const requestStream = await transport.createSendStream();
+  requestStream.write(mediaRequest);
+
+  transport.onreceivestream = (e) => {
+    const size = await readStreamUint32(e.stream);
+    if (!size || size > 512000) {
+      return;
+    }
+    const chunk = await readStreamBytes(e.stream, size);
+    if (!!chunk) {
+      sourceBuffer.appendBuffer(chunk);
+    }
+  }
+};
+
+async function readStreamUint32(stream) {
+  await stream.waitForReadable(4);
+  if (!stream.readable) {
+    return null;
+  }
+  const buffer = new Uint8Array(4);
+  const read = stream.readInto(buffer);
+  return new DataView(buffer.array).getUint32(0);
+}
+
+async function readStreamBytes(stream, count) {
+  const buffer = new ArrayBuffer(count);
+  let bufferedAmount = 0;
+  while (bufferedAmount < count) {
+    await stream.waitForReadable(1);
+    if (!stream.readable) {
+      return null;
+    }
+    const read = stream.readInto(new Uint8Array(buffer, bufferedAmount));
+    bufferedAmount += read.amount;
+  }
+  return buffer;
+}
+```
+
+## Example of receiving notifications pushed from the server, with responses
+
+```javascript
+const host = 'example.com';
+const port = 10001;
+const transport = new QuicTransport(host, port);
+
+// Note that the notifications will arrive out of order
+transport.onbidirectionalstream = (e) => {
+  const notification = readStreamUntilFin(e.stream)
+
+  if (notification) {
+    // App-specific notification encoding
+    const notificationMessage = decodeNotification(notification);
+    let n = new Notification(notificationMessage);
+    n.onclick = (e) => {
+      // App-specific click message encoding
+      const clickMessage = encodeClickMessage();
+      e.stream.write(clickMessage);
+    };
+  }
+}
+
+async function readStreamUntilFin(stream) {
+  const buffers = [];
+  let bufferedSize = 0;
+  let finished = false;
+  while (stream.readable) {
+    await stream.waitForReadable(1);
+    const buffer = new Uint8Array(stream.readableAmount);
+    const read = stream.readInto(readBuffer);
+    bufferedSize += read.amount;
+    finished = read.finished;
+  }
+  if (!finished) {
+    return null;
+  }
+
+  let joinedBuffer = new Uint8Array(bufferedSize);
+  let joinedSize = 0;
+  for (let buffer of buffers) {
+    joinedBuffer.set(buffer, joinedSize);
+    joinedSize += buffer.byteLength;
+  }
+  return joinedBuffer.array;
+}
+```
+
+## Example of requesting over HTTP and receiving media pushed out-of-order and unreliably over the same network connection
+
+```javascript
+const transport = PooledHttpTransport.getDatagramTransport();
+if (transport) {
+  await fetch('http://example.com/babyshark');
+  const datagrams = await transport.receiveDatagrams();
+  for (let data of datagrams) {
+    if (data) {
+      // Process the data
```

Comment:
Did you mean to write more here?

Reply:
Sure.  I'll just assume some hacky method like "containerizeMedia" that turns something into webm.